spec/lib/topics_bulk_action_spec.rb

```diff
@@ -158,154 +157,160 @@
+          expect do
             TopicsBulkAction.new(
-              topic.user,
+              admin,
               [topic.id],
               type: "change_category",
-              category_id: topic.category_id,
-            )
-          topic_ids = tba.perform!
-          expect(topic_ids).to be_empty
-
-          topic.reload
-          revision = topic.first_post.revisions.last
-          expect(revision).to be_nil
+              category_id: category.id,
+              silent: true,
+            ).perform!
+          end.to not_change { Notification.where(user: topic_watcher).count }
         end
-      end
-
-      context "with 'create_revision_on_bulk_topic_moves' setting disabled" do
-        before { SiteSetting.create_revision_on_bulk_topic_moves = false }
 
-        it "changes the category, doesn't create a post revision and returns the topic_id" do
-          tba =
+        it "notifies category watchers when silent is false" do
+          expect do
             TopicsBulkAction.new(
-              topic.user,
+              admin,
               [topic.id],
               type: "change_category",
               category_id: category.id,
-            )
-          topic_ids = tba.perform!
-          expect(topic_ids).to eq([topic.id])
-          topic.reload
-          expect(topic.category).to eq(category)
-
-          revision = topic.first_post.revisions.last
-          expect(revision).to be_nil
+            ).perform!
+          end.to change { Notification.where(user: category_watcher).count }.by(1)
+
+          expect(Notification.where(user: category_watcher).last.notification_type).to eq(
+            Notification.types[:watching_first_post],
+          )
         end
 
-        it "doesn't do anything when category stays the same" do
-          tba =
+        it "does not notify category watchers when silent is true" do
+          expect do
             TopicsBulkAction.new(
-              topic.user,
+              admin,
               [topic.id],
               type: "change_category",
-              category_id: topic.category_id,
-            )
-          topic_ids = tba.perform!
-          expect(topic_ids).to be_empty
+              category_id: category.id,
+              silent: true,
+            ).perform!
+          end.to not_change { Notification.where(user: category_watcher).count }
         end
       end
+
+      context "when create_revision_on_bulk_topic_moves is enabled" do
+        before { SiteSetting.create_revision_on_bulk_topic_moves = true }
+        include_examples "silent option suppresses notifications"
+      end
+
+      context "when create_revision_on_bulk_topic_moves is disabled" do
+        before { SiteSetting.create_revision_on_bulk_topic_moves = false }
+        include_examples "silent option suppresses notifications"
+      end
+    end
+
+    context "when the user can edit the topic" do
+      it "changes category and creates revision when setting enabled" do
```

Comment:
would not be better/make more sense to move these `context` to `when create_revision_on_bulk_topic_moves is enabled/disabled` context?